trecs/models/homogenizer.py

```diff
@@ -0,0 +1,221 @@
+        `num_users=200` but have `user_representation.shape` be `(300, 100)`.
+        Likewise, one cannot pass in `num_items=1000` but have
+        `item_representation.shape` be `(100, 500)`.
+    """
+
+    def __init__(  # pylint: disable-all
+        self,
+        num_users=None,
+        num_items=None,
+        num_attributes=10,
+        user_representation=None,
+        item_representation=None,
+        actual_user_representation=None,
+        actual_item_representation=None,
+        probabilistic_recommendations=False,
+        seed=None,
+        verbose=False,
+        num_items_per_iter=10,
+        homogenization_increment=0.1,
+        model_class=ContentFiltering,
+        **kwargs
+    ):
+
+        self.__class__ = type(
+            self.__class__.__name__, (model_class, object), dict(self.__class__.__dict__)
+        )
+
+        # measurements = [MSEMeasurement()]
+
+        super(self.__class__, self).__init__(
+            num_users,
+            num_items,
+            num_attributes,
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            probabilistic_recommendations,
+            seed,
+            verbose,
+            num_items_per_iter,
+            **kwargs
+        )
+
+        assert 0 < homogenization_increment < 1
+
+        gen = Generator(seed=seed)
+
+        self.homogenized_users_hat = np.tile(
+            gen.normal(size=(1, num_attributes)), (self.num_users, 1)
+        )
+        # attribute that retains the indivdiual users' preference representation in the algorithn
+        self.individual_users_hat = self.users_hat
+        # proportion of users_hat that will come from the homogenized representation. Will be updated by homogenization_increment
+        self.homogenization_proportion = 0
+        self.homogenization_increment = homogenization_increment
+
+    def _update_internal_state(self, interactions):
+        """
+        Private function that updates user profiles with data from latest
+        interactions.
+
+        Specifically, this function converts interactions into attributes.
+        For example, if user `u` has interacted with an item that has attributes
+        `a1` and `a2`, user `u`'s profile will be updated by increasing the
+        similarity to attributes `a1` and `a2`.
+
+        Parameters:
+        ------------
+            interactions: numpy.ndarray
+                An array of item indices that users have interacted with in the
+                latest step. Namely, `interactions[u]` represents the index of
+                the item that the user has interacted with.
+
+        """
+        interactions_per_user = np.zeros((self.num_users, self.num_items))
+        interactions_per_user[self.users.user_vector, interactions] = 1
+        user_attributes = np.dot(interactions_per_user, self.items_hat.T)
+
+        self.individual_users_hat += user_attributes
```

Comment:
correct me if I'm wrong, but this updates the individual user representation the same way that our ContentFiltering model does, right?